File: common/src/main/java/cc/turtl/chiselmon/api/filter/RuntimeFilter.java
package cc.turtl.chiselmon.api.filter;

import cc.turtl.turtlshell.api.core.Priority;
import com.cobblemon.mod.common.pokemon.Pokemon;
import net.minecraft.network.chat.Component;

import java.util.function.Predicate;

import static cc.turtl.chiselmon.core.util.format.ComponentUtilsKt.createComponent;

/**
 * Represents a named group of conditions specifying pokemon.
 *
 * @param id        Unique identifier for this filter (e.g., "legendaries", "custom_tiny_shinies")
 * @param name      Pretty display name string
 * @param rgb       RGB hex color used for text and glow effects involving the group.
 * @param priority  Alert priority for this filter
 * @param condition Predicate that determines if a Pokemon passes this filter
 */
public record RuntimeFilter(
        String id,
        String name,
        int rgb,
        Priority priority,
        Predicate<Pokemon> condition
) {
    public boolean matches(Pokemon pokemon) {
        return condition.test(pokemon);
    }

    public Component displayName() {
        return createComponent(name(), rgb(), false);
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/core/api/filter/RuntimeFilter.kt
package cc.turtl.chiselmon.core.api.filter

import cc.turtl.chiselmon.core.util.format.createComponent
import cc.turtl.turtlshell.api.core.Priority
import com.cobblemon.mod.common.pokemon.Pokemon
import net.minecraft.network.chat.Component
import java.util.function.Predicate

/**
 * Represents a compiled, named filter ready to match against pokemon.
 *
 * @param id        Unique identifier (e.g. "legendaries", "custom_tiny_shinies")
 * @param name      Pretty display name string
 * @param rgb       RGB color used for text and glow effects
 * @param priority  Alert priority for this filter
 * @param condition Predicate that determines if a Pokemon passes this filter
 */
data class RuntimeFilter(
    @JvmField val id: String,
    @JvmField val name: String,
    @JvmField val rgb: Int,
    @JvmField val priority: Priority,
    @JvmField val condition: Predicate<Pokemon>
) {
    fun matches(pokemon: Pokemon): Boolean = condition.test(pokemon)

    fun displayName(): Component = createComponent(name, rgb, false)
}